desc/integrals/bounce_integral.py

```diff
@@ -1,26 +1,725 @@
+        sequence of poloidal coordinates that specify the field line.
+        Likewise with θ.
+
+    """
+    if not is_reshaped:
+        B = grid.meshgrid_reshape(B, "rtz")
+
+    # Evaluating set of single variable maps is more efficient than evaluating
+    # multivariable map, so we project θ to a set of Chebyshev series.
+    T = FourierChebyshevSeries(f=theta, domain=(0, 2 * jnp.pi)).compute_cheb(alpha)
+    theta = T.evaluate(N_B)
+    xq = jnp.stack(
+        [theta, jnp.broadcast_to(cheb_pts(N_B, domain=T.domain), theta.shape)], axis=-1
+    ).reshape(*alpha.shape[:-1], alpha.shape[-1] * N_B, 2)
+    B = interp_rfft2(
+        xq=xq,
+        f=B[..., jnp.newaxis, :, :],
+        axes=(-1, -2),
+    ).reshape(*alpha.shape, N_B)
+    # Need |B| parameterized by single variable to compute roots.
+    B = PiecewiseChebyshevSeries(cheb_from_dct(dct(B, type=2, axis=-1)) / N_B, T.domain)
+    return T, B
+
+
+def _swap_pl(f):
+    # Given shape (L, num_pitch, -1) or (num_pitch, L, -1) or (num_pitch, -1)
+    # swap L and num_pitch axes.
+    assert f.ndim <= 3
+    return jnp.swapaxes(f, 0, -2)
+
+
+# TODO: After GitHub issue #1034 is resolved, we should pass in the previous
+#  θ(α, ζ) coordinates as an initial guess for the next coordinate mapping.
+#  Perhaps tell the optimizer to perturb the coefficients of the
+#  θ(α, ζ) directly? think this is equivalent to perturbing lambda.
+
+
+class Bounce2D(IOAble):
+    """Computes bounce integrals using two-dimensional pseudo-spectral methods.
+
+    The bounce integral is defined as ∫ f(λ, ℓ) dℓ, where
+        dℓ parameterizes the distance along the field line in meters,
+        f(λ, ℓ) is the quantity to integrate along the field line,
+        and the boundaries of the integral are bounce points ℓ₁, ℓ₂ s.t. λ|B|(ℓᵢ) = 1,
+        where λ is a constant defining the integral proportional to the magnetic moment
+        over energy and |B| is the norm of the magnetic field.
+
+    For a particle with fixed λ, bounce points are defined to be the location on the
+    field line such that the particle's velocity parallel to the magnetic field is zero.
+    The bounce integral is defined up to a sign. We choose the sign that corresponds to
+    the particle's guiding center trajectory traveling in the direction of increasing
+    field-line-following coordinate ζ.
+
+    Notes
+    -----
+    Brief description of algorithm.
+
+    Magnetic field line with label α, defined by B = ∇ρ × ∇α, is determined from
+        α : ρ, θ, ζ ↦ θ + λ(ρ,θ,ζ) − ι(ρ) [ζ + ω(ρ,θ,ζ)]
+    Interpolate Fourier-Chebyshev series to DESC poloidal coordinate.
+        θ : α, ζ ↦ tₘₙ exp(jmα) Tₙ(ζ)
+      |B| : α, ζ ↦  bₙ(θ(α, ζ)) Tₙ(ζ)
+    Compute bounce points.
+      r(ζₖ) = |B|(ζₖ) − 1/λ = 0
+    Interpolate smooth components of integrand with FFTs.
+      G : α, ζ ↦ gₘₙ exp(j [m θ(α,ζ) + n ζ] )
+    Perform Gaussian quadrature after removing singularities.
+      Fᵢ : λ, ζ₁, ζ₂ ↦  ∫ᵢ f(λ, ζ, {Gⱼ}) dζ
```

Comment:
I chose quadrature over Runge-Kutta method of the form ∂Fᵢ/∂ζ =f(λ, ζ, {Gⱼ}) with boundary conditions Fᵢ(ζ₁) = 0.  I did so because a fourth order Runge-kutta method is equivalent to a quadrature with Simpson's rule, and our quadratures resolve these integrals much more efficiently than Simpson's rule.

FYI, Neo uses a fourth order Runge-Kutta. That method would fail for strongly singular integrals such as the ones in the Gammas. Stellopt uses tanh-sinh quadrature for that, which we tested performed poorly in our implementation.